archive/100-theorems-list/88_derangements_formula.lean

```diff
@@ -0,0 +1,468 @@
+which gives an explicit formula for the number of derangements (permutations without fixed
+points) on `n` elements.
+
+The core of the proof is an equivalence between "derangements on `n+2` elements", and
+an expression involving derangements on `n` and `n+1` elements. This gives a recurrence
+relation `d(n+2) = (n+1) * (d(n) + d(n+1))`, from which a sum-based formula can be derived.
+-/
+
+open equiv fintype
+open_locale big_operators
+
+section definitions
+
+/-- A permutation is a derangement if it has no fixed points. -/
+def is_derangement {α : Type*} (f : perm α) : Prop := (∀ x : α, f x ≠ x)
+
+/-- The set of derangements on `α`. -/
+def derangements (α : Type*) : set (perm α) := {f : perm α | is_derangement f}
+
+end definitions
+
+
+section derangement_lemmas
+
+/-- If `α` is equivalent to `β`, then `derangements α` is equivalent to `derangements β`. -/
+def derangement_congr {α β : Type*} (e : α ≃ β) : (derangements α ≃ derangements β) :=
+begin
+  refine subtype_equiv (perm_congr e) _,
+  intro f,
+  unfold derangements is_derangement,
+  simp,
+  rw ← not_iff_not,
+  push_neg,
+  split,
+  { rintro ⟨x, hx_fixed⟩,
+    use e x,
+    simp [hx_fixed] },
+  { rintro ⟨y, hy_fixed⟩,
+    use e.symm y,
+    apply e.injective,
+    simp [hy_fixed] }
+end
+
+end derangement_lemmas
+
+
+section ignore_swap
+
+variables {α : Type*} [decidable_eq α]
+
+/-- A predicate describing if permutation `f` swaps `a` and `b`.
+    Note that `a` and `b` need not be distinct. -/
+def swaps {α : Type*} (f : perm α) (a b : α) : Prop := f a = b ∧ f b = a
+
+/-- The set of permutations that swap `a` and `b` is equivalent to permutations on all other
+    elements. -/
+def perm_ignore_swap (a b : α) : { f : perm α | swaps f a b } ≃ perm ({a, b}ᶜ : set α) :=
+begin
+  let a_and_b : set α := {a, b},
+
+  -- TODO why can't it infer this is decidable?
+  haveI : decidable_pred a_and_b := by
+  { dsimp [decidable_pred, a_and_b],
+    intro x,
+    apply set.decidable_mem },
+
+  let swap : perm α := equiv.swap a b,
+  let swap' : perm a_and_b := perm.subtype_perm swap (by {
+    intro x,
+
+    simp,
+    conv_rhs { rw apply_eq_iff_eq_symm_apply, rw apply_eq_iff_eq_symm_apply },
+    simp [swap, or_comm] }),
+
+  transitivity {f : perm α // ∀ x : a_and_b, f x = swap' x},
+  { apply subtype_equiv_right,
+    intro σ,
+    simp [swaps] },
+  { exact equiv.set.compl swap' },
+end
+
+-- TODO should this be a simp lemma?
+lemma perm_ignore_two_cycle_eq (a b : α) (f : perm α) (hf : swaps f a b) (c : ({a, b}ᶜ : set α)):
+  ((perm_ignore_swap a b) ⟨f, hf⟩ c : α) = f c := by simp [perm_ignore_swap, equiv.set.compl]
+
+-- TODO this name definitely isn't up to snuff... but what would it be called?
+lemma perm_ignore_two_cycle_derangement_iff (a b : α) (hab : a ≠ b) (f : perm α) (hf : swaps f a b) :
+  is_derangement f ↔ is_derangement ((perm_ignore_swap a b) ⟨f, hf⟩) :=
+begin
+  split,
+
```

Comment:
Usually don't leave blank spaces in proof environments. It's okay, but even better is to fill that blank space with a helpful comment. :-)